stan/math/prim/mat/prob/neg_binomial_2_log_glm_lpmf.hpp

```diff
@@ -0,0 +1,195 @@
+
+      using std::exp;
+      using Eigen::Dynamic;
+      using Eigen::Matrix;
+      using Eigen::Array;
+
+      if (!(stan::length(n) && stan::length(x) && stan::length(beta)
+            && stan::length(phi)))
+        return 0.0;
+
+      T_partials_return logp(0.0);
+
+      check_nonnegative(function, "Failures variables", n);
+      check_finite(function, "Matrix of independent variables", x);
+      check_finite(function, "Weight vector", beta);
+      check_finite(function, "Intercept", alpha);
+      check_positive_finite(function, "Precision parameter", phi);
+      check_consistent_sizes(function,
+                             "Rows in matrix of independent variables",
+                             x.col(0), "Vector of dependent variables",  n);
+      check_consistent_sizes(function,
+                             "Rows in matrix of independent variables",
+                             x.col(0), "Vector of failure counts",  phi);
+      check_consistent_sizes(function,
+                             "Columns in matrix of independent variables",
+                             x.row(0), "Weight vector",  beta);
+
+      if (!include_summand<propto, T_x, T_beta, T_alpha, T_precision>::value)
+        return 0.0;
+
+      const size_t N = x.col(0).size();
+      const size_t M = x.row(0).size();
+
+      Array<T_partials_return, Dynamic, 1> n_arr(N, 1);
+      Array<T_partials_return, Dynamic, 1> phi_arr(N, 1);
+      {
+        scalar_seq_view<T_n> n_vec(n);
+        scalar_seq_view<T_precision> phi_vec(phi);
+        for (size_t n = 0; n < N; ++n) {
+          n_arr[n] = n_vec[n];
+          phi_arr[n] = value_of(phi_vec[n]);
+        }
+      }
+
+      Matrix<T_partials_return, Dynamic, 1> beta_dbl(M, 1);
+      {
+        scalar_seq_view<T_beta> beta_vec(beta);
+        for (size_t m = 0; m < M; ++m) {
+          beta_dbl[m] = value_of(beta_vec[m]);
+        }
+      }
+      Matrix<T_partials_return, Dynamic, Dynamic> x_dbl = value_of(x);
+
+      Array<T_partials_return, Dynamic, 1> theta_dbl = (x_dbl * beta_dbl
+      + Matrix<double, Dynamic, 1>::Ones(N, 1) * value_of(alpha)).array();
+      Array<T_partials_return, Dynamic, 1> log_phi = phi_arr.log();
+      Array<T_partials_return, Dynamic, 1> logsumexp_eta_logphi =
+        theta_dbl.binaryExpr(log_phi,
+                             [](T_partials_return xx, T_partials_return yy) {
+                               return log_sum_exp(xx, yy);
+                             });
+      Array<T_partials_return, Dynamic, 1> n_plus_phi = n_arr + phi_arr;
+
+      // Compute the log-density.
+      if (include_summand<propto>::value) {
+        logp -= (n_arr
+          + Array<double, Dynamic, 1>::Ones(N, 1))
+          .unaryExpr([](T_partials_return xx) {
+                       return lgamma(xx);
+                     }).sum();
+      }
+      if (include_summand<propto, T_precision>::value) {
+        logp += (phi_arr.binaryExpr(phi_arr, [](T_partials_return xx,
```

Comment:
A thing we might prefer to do here is define `multiply_log` and `lgamma` (below) for Eigen types - it's possible we could save a fair bit of computation by doing that intelligently. This version I think will create a lot of intermediate `var`s. @bob-carpenter Am I right about that? What do you think about this `binaryExpr` (basically `map` + `zip` for Eigen Arrays - runs the function coefficient-wise along both of them in lock step, returning the result at each point as a new array) being here in this file vs. as a new specialization of multiply_log? I'm guessing the only real reason to do the latter would be potential performance gains, and to maybe save a slight amount of typing.

Reply:
Yes, that is what `binaryExpr` is supposed to do, `map` after `zip`, I think. I was hoping that this would already be reasonably efficient. I now see that  `apply_scalar_unary` in the Stan math library does something similar to `unaryExpr`. Would that be more efficient, @bob-carpenter ? Also, would you have a suggested way of implementing something like `binaryExpr`, perhaps using `apply_scalar_unary` and a `zip`?

Reply:
I mean this version is definitely good enough; I'm assuming you did the performance test and the custom derivatives were already better than autodiffing through. If someone (Bob? @bgoodri?) thinks there will be more performance gains in a custom version we can always add that later.

Reply:
Yes, this version was 4.5x faster than the composite of existing primitives, in my tests. I should probably still do  more extensive tests though on different shapes of data to be sure of how robust the speedup of the new primitives is.